File: 백준/Gold/2230. 수 고르기/수 고르기.py
import sys
# sys.stdin = open("input.txt", "r")

N, M = map(int, input().split())
arr = list()
for _ in range(N):
    arr.append(int(input()))
arr.sort()

p, q = 0, 0
ans = sys.maxsize
while q < N and p < N:
    tmp = arr[q] - arr[p]
    if tmp < M:
        q += 1
    else:
        p += 1
        ans = min(ans, tmp)

print(ans)

File: 백준/Gold/2230. 수 고르기/수 고르기.cc
#include <bits/stdc++.h>
using namespace std;

vector<long long> arr(100001);
int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    int N; cin >> N ;
    long long M; cin >> M;
    for (int n=0; n<N; n++) cin >> arr[n];

    sort(arr.begin(), arr.begin()+N);
    int ep = 0;
    long long ans = 2000000001;
    for (int sp=0; sp<N; sp++)
    {
        while (ep < N && arr[ep] - arr[sp] < M) ep++;
        if (ep >= N) break;
        ans = min({ans, arr[ep] - arr[sp]});
        if (ans == M)
        {
            cout << M; return 0;
        }
    }
    cout << ans;
    return 0;
}